File: java/3655.xor-after-range-multiplication-queries-ii.java
import java.util.*;

class Solution {
    private static final int MOD = (int) 1e9 + 7;

    // Fast exponentiation: x^y mod MOD (binary exponentiation)
    private long pow(int x, int y) {
        long res = 1;
        long base = x % MOD;

        while (y > 0) {
            if ((y & 1) == 1) {  // Java: explicit comparison for bitwise result
                res = (res * base) % MOD;
            }
            base = (base * base) % MOD;
            y >>= 1;
        }
        return res;
    }

    // Modular inverse using Fermat's Little Theorem: x^(-1) ≡ x^(MOD-2) mod MOD
    private long inv(int x) {
        return pow(x, MOD - 2);
    }

    public int xorAfterQueries(int[] nums, int[][] queries) {
        final int n = nums.length;
        final int limit = (int) Math.sqrt(n) + 1;

        // Map: step size k → list of queries with that step (dense queries)
        Map<Integer, List<int[]>> dense = new HashMap<>();

        // q[0]=left, q[1]=right, q[2]=step, q[3]=value
        for (int[] q : queries) {
            if (q[2] >= limit) {
                // Sparse query: apply immediately (O(n/k) per query)
                for (int i = q[0]; i <= q[1]; i += q[2]) {
                    nums[i] = (int) ((1L * nums[i] * q[3]) % MOD);
                }
            } else {
                // Dense query: batch by step size for later processing
                dense.computeIfAbsent(q[2], k -> new ArrayList<>()).add(q);
            }
        }

        // Process batched dense queries using multiplicative difference array
        for (Map.Entry<Integer, List<int[]>> entry : dense.entrySet()) {
            int k = entry.getKey();
            List<int[]> qlist = entry.getValue();

            // mul[i] = cumulative multiplier to apply at index i
            int[] mul = new int[n];
            Arrays.fill(mul, 1);

            for (int[] q : qlist) {
                // Left bound: multiply by value
                mul[q[0]] = (int) ((1L * mul[q[0]] * q[3]) % MOD);

                // Right bound + step: divide by value (via modular inverse)
                int steps = (q[1] - q[0]) / k;
                int next = q[0] + (steps + 1) * k;
                if (next < n) {
                    mul[next] = (int) ((1L * mul[next] * inv(q[3])) % MOD);
                }
            }

            // Propagate multipliers along arithmetic progression with step k
            for (int i = 0; i < n; i++) {
                if (i >= k) {
                    mul[i] = (int) ((1L * mul[i] * mul[i - k]) % MOD);
                }
                nums[i] = (int) ((1L * nums[i] * mul[i]) % MOD);
            }
        }

        // Compute XOR of all elements (equivalent to std::accumulate with bit_xor)
        int result = 0;
        for (int num : nums) {
            result ^= num;
        }
        return result;
    }
}

File: cpp/3655.xor-after-range-multiplication-queries-ii.cpp
#include <cmath>
#include <functional> // std::bit_xor
#include <numeric>    // std::accumulate
#include <unordered_map>
#include <vector>

using std::vector, std::unordered_map;

class Solution {
private:
    static const int MOD = int(1e9) + 7;

    // exponentiate x ^ y modulo MOD
    long long pow(int x, int y) {
        long long res = 1;
        long long base = x % MOD;

        while (y > 0) {
            if (y & 1) {
                res = (res * base) % MOD;
            }

            base = (base * base) % MOD;
            y >>= 1;
        }

        return res;
    }

    // invert x => 1 / x modulo MOD
    // Fermat little theoreme:
    //        x ^ (p - 1) = 1 mod p
    // Divide both sides by x:
    //        x ^ (p - 2) = 1 / x
    long long inv(int x) { return pow(x, MOD - 2); }

public:
    int xorAfterQueries(vector<int>& nums, vector<vector<int>>& queries) {
        const int n = nums.size();
        const int limit = std::sqrt(n) + 1;
        unordered_map<int, vector<vector<int>>> dense; // group by step size

        // q(0 - left, 1 - right, 2 - step, 3 - value)
        for (auto& q : queries) {
            if (q[2] >= limit) { // if sparse
                // apply adhoc: multiply by value each number left to right
                // (inclusive) with step
                for (int i = q[0]; i <= q[1]; i += q[2]) {
                    nums[i] = (1LL * nums[i] * q[3]) % MOD;
                }
            } else { // if dense
                // save for later grouping by step size
                dense[q[2]].push_back(q);
            }
        }


        // apply dense rows according to step size k
        for (auto& [k, qlist] : dense) {
            // multiply vector;
            vector<int> mul(n, 1);
            
            for (auto& q : qlist) {
                // left bound multiply
                mul[q[0]] = (1LL * mul[q[0]] * q[3]) % MOD;
                // next to right bound
                int steps = (q[1] - q[0]) / k;     // number of steps
                int next = q[0] + (steps + 1) * k; // next step index

                // right bound divide by (if withing a bounds)
                if (next < n) {
                    mul[next] = (1LL * mul[next] * inv(q[3])) % MOD;
                }
            }

            for (int i = 0; i < n; i++) {
                if (i >= k) {
                    mul[i] = (1LL * mul[i] * mul[i - k]) % MOD;
                }
                nums[i] = (1LL * nums[i] * mul[i]) % MOD;
            }
        }

        return std::accumulate(nums.begin(), nums.end(), 0,
                               std::bit_xor<int>());
    }
};